### PR TITLE
Adgudime/apex sac

### PR DESCRIPTION
**Pull request by AdityaGudimella**
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: rllib/utils/exploration/per_worker_epsilon_greedy.py [hunk 1]
```diff
@@ -54,10 +54,12 @@ def reset_schedule(self,
                        epsilon_schedule=None
                        ):
         epsilon_schedule = None
-        if self.num_workers > 0:
+        if self.num_workers == 1:
```
Comment (RuofanKong) [inline rllib/utils/exploration/per_worker_epsilon_greedy.py:57]: Just curious -- what would happen if we don't add this condition:)

File: rllib/agents/ddpg/apex.py [hunk 2]
```diff
@@ -16,7 +16,7 @@
         "n_step": 3,
         "num_gpus": 0,
         "num_workers": 32,
-        "buffer_size": 2000000,
+        "buffer_size": 2000001,
```
Comment (RuofanKong) [inline rllib/agents/ddpg/apex.py:19]: Just curious - what would happen if we don't make the batch size change?

Comment (AdityaGudimella) [reply rllib/agents/ddpg/apex.py:19]: Nothing. I honestly have no idea how this got in here.

File: rllib/utils/exploration/per_worker_epsilon_greedy.py (hunk 1 above)
Comment (AdityaGudimella) [reply rllib/utils/exploration/per_worker_epsilon_greedy.py:57]: We get a zero division error if we set num_workers to 1. So I had to handle that condition separately.

**Comment by AdityaGudimella** [comment]
> Thank you @AdityaGudimella for the work!! One thing that I try to understand is the reason for reshape action (for both dqn_tf_policy and async_replay_optimizer). I also have the related concern on if changing dqn_tf_policy may impact apex-dqn. probably I can learn more from you on an offline sync. Otherwise, LGTM.

I have tests for apex-dqn and dqn and the change did indeed affect them. Which is why the check for Box action space / discrete action space is present. The async replay optimizer seems to have only been built for DQN and so they didn't support Box space (just an assumption because I don't know what they original person intended without any comments in the code base...). I'd be happy to chat offline with you about this if you wish. :)

**Comment by RuofanKong** [comment]
> > Thank you @AdityaGudimella for the work!! One thing that I try to understand is the reason for reshape action (for both dqn_tf_policy and async_replay_optimizer). I also have the related concern on if changing dqn_tf_policy may impact apex-dqn. probably I can learn more from you on an offline sync. Otherwise, LGTM.
> 
> I have tests for apex-dqn and dqn and the change did indeed affect them. Which is why the check for Box action space / discrete action space is present. The async replay optimizer seems to have only been built for DQN and so they didn't support Box space (just an assumption because I don't know what they original person intended without any comments in the code base...). I'd be happy to chat offline with you about this if you wish. :)

I see. Let me think about it. I'm asking as Apex-DDPG uses the async replay optimizer with box action type as well. Changing it might break it (Ray Fork tests may reflect on it), although we don't use it for now.

File: rllib/agents/dqn/dqn_tf_policy.py [hunk 3]
```diff
@@ -369,8 +373,11 @@ def postprocess_nstep_and_prio(policy, batch, other_agent=None, episode=None):
 
     # Prioritize on the worker side
     if batch.count > 0 and policy.config["worker_side_prioritization"]:
+        actions = batch[SampleBatch.ACTIONS]
```
Comment (Edilmo) [inline rllib/agents/dqn/dqn_tf_policy.py:376]: Why do we need the "flattening" here?

I understand that RLlib requires a vector of actions, but that is a known limitation that is true for all  algorithms, isn't it?

File: rllib/optimizers/async_replay_optimizer.py [hunk 4]
```diff
@@ -467,6 +469,25 @@ def step(self):
...
+                            ) is not None,
+                            isinstance(
+                                self.local_worker.policy_map[
+                                    "default_policy"
+                                ].action_space,
+                                gym.spaces.Box
+                            )
+                        ))
+                        if is_box_action_space:
+                            batch["actions"] = batch["actions"].reshape((-1, 1))
```
Comment (Edilmo) [inline rllib/optimizers/async_replay_optimizer.py:490]: Same questions here, why the reshaping?

File: rllib/agents/ddpg/apex.py (hunk 2 above)
Comment (AdityaGudimella) [reply rllib/agents/ddpg/apex.py]: Removed

File: rllib/agents/trainer_factory.py [hunk 5]
```diff
@@ -4,11 +4,12 @@
...
-from ray.rllib.agents.ddpg import apex, td3
-from ray.rllib.agents.dqn import apex
-from ray.rllib.agents.ppo import ddppo, appo
-from ray.rllib.agents.trainer import Trainer
+from rllib.agents import ddpg, dqn, ppo, sac
+from rllib.agents.ddpg import apex, td3
+from rllib.agents.dqn import apex
+from rllib.agents.ppo import ddppo, appo
+from rllib.agents.sac import apex
+from rllib.agents.trainer import Trainer
```
[the remark this answers is not in the document]
Comment (AdityaGudimella) [reply rllib/agents/trainer_factory.py]: Changed.

File: rllib/tests/agents/test_learning.py [hunk 6]
```diff
@@ -20,6 +20,7 @@
 
 logger = logging.getLogger(__name__)
 
+
```
[the remark this answers is not in the document]
Comment (AdityaGudimella) [reply rllib/tests/agents/test_learning.py:19]: Removed

**Pull request merged**